lib/api/api_auth_webauthn.go

```diff
@@ -0,0 +1,470 @@
+// Copyright (C) 2023 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package api
+
+import (
+	"bytes"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"reflect"
+	"slices"
+	"time"
+
+	webauthnProtocol "github.com/go-webauthn/webauthn/protocol"
+	webauthnLib "github.com/go-webauthn/webauthn/webauthn"
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/db"
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/sliceutil"
+)
+
+func newWebauthnEngine(cfg config.Wrapper) (*webauthnLib.WebAuthn, error) {
+	guiCfg := cfg.GUI()
+
+	displayName := "Syncthing"
+	if dev, ok := cfg.Device(cfg.MyID()); ok && dev.Name != "" {
+		displayName = "Syncthing @ " + dev.Name
+	}
+
+	rpId := guiCfg.WebauthnRpId
+	if rpId == "" {
+		guiCfgStruct := reflect.TypeOf(guiCfg)
+		field, found := guiCfgStruct.FieldByName("WebauthnRpId")
+		if !found {
+			return nil, fmt.Errorf(`Field "WebauthnRpId" not found in struct GUIConfiguration`)
+		}
+		rpId = field.Tag.Get("default")
+		if rpId == "" {
+			return nil, fmt.Errorf(`Default tag not found on field "WebauthnRpId" in struct GUIConfiguration`)
+		}
+	}
+
+	origin := guiCfg.WebauthnOrigin
+	if origin == "" {
+		guiCfgStruct := reflect.TypeOf(guiCfg)
+		field, found := guiCfgStruct.FieldByName("WebauthnOrigin")
+		if !found {
+			return nil, fmt.Errorf(`Field "WebauthnOrigin" not found in struct GUIConfiguration`)
+		}
+		origin = field.Tag.Get("default")
+		if origin == "" {
+			return nil, fmt.Errorf(`Default tag not found on field "WebauthnOrigin" in struct GUIConfiguration`)
+		}
+	}
+
+	return webauthnLib.New(&webauthnLib.Config{
+		RPDisplayName: displayName,
+		RPID:          rpId,
+		RPOrigins:     []string{origin},
+	})
+}
+
+type webauthnService struct {
+	tokenCookieManager             *tokenCookieManager
+	miscDB                         *db.NamespacedKV
+	miscDBKey                      string
+	engine                         *webauthnLib.WebAuthn
+	cfg                            config.Wrapper
+	evLogger                       events.Logger
+	userHandle                     []byte
+	registrationState              webauthnLib.SessionData
+	authenticationState            webauthnLib.SessionData
+	credentialsPendingRegistration []config.WebauthnCredential
+}
+
+func newWebauthnService(tokenCookieManager *tokenCookieManager, cfg config.Wrapper, evLogger events.Logger, miscDB *db.NamespacedKV, miscDBKey string) (webauthnService, error) {
+	engine, err := newWebauthnEngine(cfg)
+	if err != nil {
+		return webauthnService{}, err
+	}
+
+	userHandle, err := base64.URLEncoding.DecodeString(cfg.GUI().WebauthnUserId)
+	if err != nil {
+		return webauthnService{}, err
+	}
```

Comment:
<details><summary>nit</summary>

My goal: Avoid the need to return an error.

I added comments in `newWebauthnEngine` why I think the code that may produce an error there isn't needed.

For the base64 decoding I'd like to move that into lib/config: On every change to config we run checks that the config is valid. Lets check that the string is valid base64 there, and fail the update otherwise. And then have a method `DecodedWebauthnUserId() string` (no error) on `GUIConfiguration`. Resp. given the config value itself is supposed to be base64, we could indicate that in the name: Config value named `WebauthnUserIdBase64` and then the method is just `WebauthnUserId`.

</details>

Reply:
Sure. Could you help me find where this validation is performed?

Reply:
Actually, on further thought, we don't actually need to decode this at all - we can just make the field `bytes` in the proto definition and take the raw value. That gets a bit messy in `config.xml` with fully-random data, but we don't need every bit to be random. We can just take some random byte string, encode it in printable characters and then use that encoding as the binary value.

If that seems reasonable, this is addressed by https://github.com/emlun/syncthing/pull/6.

Reply:
Fixed by https://github.com/emlun/syncthing/pull/6.